File: machine learning models/Logistic Regression Bankruptcy Prediction/Python/bankruptcylg.py
import lreg_bankrupt_lib as lib
import csv
import numpy as np
import random
from scipy.optimize import minimize

import numpy as np
import itertools

def computeCost(theta, X, y, lambdad=0):
    m = len(y)                  # number of training examples
    z = np.dot(X, theta)
    h = sigmoid(z)
    # define variable theta0 to spare intercept from regularization
    theta0 = theta
    theta0[0] = 0
    cost = 1 / m * sum(- y * np.log(h) - (1 - y) * np.log(1 - h)) + \
        lambdad / m / 2 * sum(theta0 ** 2)
    return cost

# function computing the gradient of cost function
def cost_grad(theta, X, y, lambdad=0):
    m = len(y)                  # number of training examples
    z = np.dot(X, theta)
    h = sigmoid(z)
    # define variable theta0 to spare intercept from regularization
    theta0 = theta
    theta0[0] = 0
    grad = (1 / m * np.dot(h - y, X)) + lambdad * theta0 / m
    return grad

# construct polynomial terms from original predictors with order restrictted
def mapping(X, degree):
    m = np.size(X, 0)
    n = np.size(X, 1)
    power = selectk(range(degree + 1), n)  # construct degree grid
    ind = np.sum(power, axis=1) <= degree  # restrict order
    power = power[ind, :]
    p = np.size(power, 0)
    Xmap = np.ones([m, p + 1])  # construct polynomial terms
    for i in range(p):
        aterm = np.ones(m)
        for j in range(n):
            aterm = aterm * X[:, j] ** power[i, j]
        Xmap[:, i + 1] = aterm
    return Xmap

# function that normalizes each predictor
def normalize(X):
    m = np.size(X, 0)
    mu = np.mean(X, 0)           # column mean
    sigma = np.std(X, 0, ddof=1)  # set divisor to be m - 1
    muMat = np.tile(mu, (m, 1))
    sigmaMat = np.tile(sigma, (m, 1))
    Xnorm = np.divide(X - muMat, sigmaMat)
    return Xnorm

# function generating grid based on v of dimension k
def selectk(v, k):
    grid = list(itertools.product(v, repeat=k))
    return np.array(grid)

# function simply computing response
def sigmoid(z):
    g = np.ones(np.size(z)) / (1 + np.exp(- z))
    return g

# load data
datPath = r"C:\xxx\bankruptcy.csv"

# read data from csv file
with open(datPath, 'rb') as csvfile:
    csvData = csv.reader(csvfile)
    datList = []
    for row in csvData:
        datList.append(row[1:])  # exclude the first char column, Firm

# get the colnames in the first row and remove it
colNames = datList.pop(0)

# convert list to matrix
data = np.array(datList, dtype="float")

# randomly select 70% data points as training set
nRow = np.size(data, 0)
trainInd = random.sample(range(nRow), int(0.7 * nRow))
testInd = np.ones(nRow, np.bool)
testInd[trainInd] = 0

# set up design matrix, response, and initial guess for training data
X = data[trainInd, 0: 12]
X = lib.normalize(X)           # standardize data column-wise
y = data[trainInd, 12]
m = np.size(X, 0)               # number of rows
n = np.size(X, 1)               # number of columns
X = np.concatenate((np.tile(1, (m, 1)), X), 1)  # add intercept
theta = np.zeros(n + 1)

# set up test set
testX = data[testInd, 0: 12]
testX = lib.normalize(testX)
m = np.size(testX, 0)               # number of rows
testX = np.concatenate((np.tile(1, (m, 1)), testX), 1)  # add intercept
testY = data[testInd, 12]

# No regularization
# Optimization using scipy.optimize.minimize
res1 = minimize(lib.computeCost, theta, args=(X, y, ),jac=lib.cost_grad, options={"maxiter": 5000})

# Accuracy with training set
pred = lib.sigmoid(np.dot(testX, res1.x)) >= 0.5
print("Accuracy: %f\n" % (np.mean(pred == testY) * 100))

# Now use regularization
# set up design matrix again ithout intercept as training set
X = data[trainInd, 0: 12]
X = lib.normalize(X)           # standardize data column-wise

# Mapping to higher dimensional space
X2 = lib.mapping(X, 2)         # rename to avoid overwrite more than onece
theta2 = np.zeros(np.size(X2, 1))

# set up test set
testX = data[testInd, 0: 12]
testX = lib.normalize(testX)   # standardize data column-wise

# Mapping to higher dimensional space
testX2 = lib.mapping(testX, 2)  # rename to avoid overwrite more than onece

lambdad = 0.01

# Optimization using scipy.optimize.minimize
res2 = minimize(lib.computeCost, theta2, args=(X2, y, lambdad, ),jac=lib.cost_grad, options={"maxiter": 5000})

# Compute accuracy on our training set
pred2 = lib.sigmoid(np.dot(testX2, res2.x)) >= 0.5
print("Accuracy: %f\n" % (np.mean(pred2 == testY) * 100))

File: machine learning models/Logistic Regression Bankruptcy Prediction/MATLAB/bankruptcylg.m
function main()
	%% ================== Reset the workspace ===================
	clear all; close all; clc;
	%% ================== No regularization =====================
	[data,txt,raw] = xlsread('bankruptcy.xls','Sheet1');
	X = data(:,1:12);
	X = normalize(X);
	y = data(:,13);                                                                                                                                                                                                                                                                                                                                                                                                                                        
	[m,n] = size(X);
	X = [ones(m,1) X];
	theta = zeros(n+1,1);

	% Optimization using fminunc 
	options = optimset('GradObj', 'on', 'MaxIter', 100);
	[theta,cost] = fminunc(@(t)(computeCost(t,X,y)),theta,options);

	% Accuracy with training set
	pred = sigmoid(X*theta) >= 0.5;
	fprintf('Accuracy: %f\n', mean(double(pred==y))*100);

	%% ============= Now use regularization  ====================
	% Reload the data
	[data,txt,raw] = xlsread('bankruptcy.xls','Sheet1');
	X = data(:,1:12);
	X = normalize(X);
	y = data(:,13);

	% Mapping to higher dimensional space
	X = mapping(X,2);
	theta = zeros(size(X,2),1);
	lambda = 0.01;

	% Optimization using fminunc
	options = optimset('GradObj','on','MaxIter',50);
	[theta,cost] = fminunc(@(t)(computeCost(t,X,y,lambda)),theta,options);

	% Compute accuracy on our training set
	pred = sigmoid(X*theta) >= 0.5;
	fprintf('Accuracy: %f\n', mean(double(pred==y))*100);

	function [cost,grad] = computeCost(theta,X,y,lambda)
		m = length(y); % number of training examples
		cost = 0;
		grad = zeros(size(theta));
		if nargin < 4
			lambda = 0; % this lambda=0 make the reularized term to go away, when a value is passed it is the step
		end
		z = X*theta; % z is m x d
		h = sigmoid(z); % same as z
		grad = (1/m * (h-y)' * X) + lambda * [0;theta(2:end)]'/m; % grad is 1 x d, theta(1) - meaning theta_0 should not be regularized
		cost =  1/(m) * sum(-y .* log(h) - (1-y) .* log(1-h)) + lambda/m/2*sum(theta(2:end).^2); % theta(1) - meaning theta_0 should not be regularized
	end

	function Xmap = mapping(X,degree)
		[m,n] = size(X); % m is number of variables
		power = selectk(0:degree,n);
		ind = (sum(power,2)<=degree & sum(power,2)>0);
		power = power(ind,:);
		p = size(power,1); % number of  polynomial terms
		Xmap = ones(m,1);
		for i = 1:p
			aterm = ones(m,1);
			for j = 1:n
				aterm = aterm .* X(:,j).^power(i,j);    
			end
			Xmap(:,end+1) = aterm;
		end
	end
	function [Xnorm] = normalize(X)
		[m,n] = size(X);
		mu = mean(X);
		sigma = std(X);
		Xnorm = (X-repmat(mu,m,1))./repmat(sigma,m,1);
	end
	function [pca,w,k]=pca(data,para)
		[n,d]=size(data);
		covm=cov(data);
		[E,D]=eig(covm); % can also use pca functions in Matlab for principal components
		[val,loc]=sort(diag(D),'descend');
		if para>=1
			k=para;
		else
			k=sum((cumsum(val)/sum(val))<=para);
			if k==0 k=1; end;
		end
		E=E(:,loc);
		w=E(:,1:k);
		length=sqrt(sum(w.^2));
		w=w./(ones(d,1)*length);
		pca= data*w; 
		% pca=(data - repmat(mean(data),n,1))*w; % remove mean before projection
	end
	function y = selectk(v,k)
		m = length(v);
		X = cell(1,k);
		[X{:}] = ndgrid(v);
		X = X(end:-1:1);
		y = cat(k+1,X{:});
		y = reshape(y,[m^k,k]);
	end
	function g = sigmoid(z)
		g = ones(size(z))./(1+exp(-z));
	end	
end